### PR TITLE
Fix cuda detection script

### LINKED ISSUE #22507: Downloading MacOSX build from source has error on last step: [Errno2 ] No such file or directory
### Please note that this issue tracker is not a help form and this issue will be closed.

We have a set of [listed resources available on the website](https://pytorch.org/resources). Our primary means of support is our discussion forum:

- [Discussion Forum](https://discuss.pytorch.org/)

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes https://github.com/pytorch/pytorch/issues/22507

### CODE REVIEW

File: tools/setup_helpers/cuda.py [hunk 1]
```diff
@@ -11,10 +11,13 @@
...
-        proc = Popen(['where', 'nvcc.exe'], stdout=PIPE, stderr=PIPE)
-    else:
-        proc = Popen(['which', 'nvcc'], stdout=PIPE, stderr=PIPE)
+    try:
+        if IS_WINDOWS:
+            proc = Popen(['where', 'nvcc.exe'], stdout=PIPE, stderr=PIPE)
+        else:
+            proc = Popen(['which', 'nvcc'], stdout=PIPE, stderr=PIPE)
+    except Exception:
+        return None
```
Comment (xuhdev) [inline tools/setup_helpers/cuda.py]: You can consider copying out the which utility to `__init__.py`:

https://github.com/pytorch/pytorch/blob/a6441c00d663892e5b1d2fc2e07bb8ad9f4244a0/tools/setup_helpers/cmake.py#L25

Comment (peterjc123) [reply tools/setup_helpers/cuda.py]: Usually, the `__init__.py` should remain empty, what about something like `common.py` or `utils.py`?

Comment (peterjc123) [reply tools/setup_helpers/cuda.py]: Oh, there is `env.py`. I think I'll move it there.

**Comment by peterjc123** [comment]
@xuhdev Changed according to your suggestion.

File: tools/setup_helpers/cuda.py [hunk 2]
```diff
@@ -4,27 +4,19 @@
...
+    nvcc = which('nvcc')
+    if nvcc is not None:
         if IS_WINDOWS:
-            if out.find('\r\n') != -1:
-                out = out.split('\r\n')[0]
-            out = os.path.abspath(os.path.join(os.path.dirname(out), ".."))
-            out = out.replace('\\', '/')
-            out = str(out)
-        return os.path.dirname(out)
+            nvcc = nvcc.replace('\\', '/')
```
Comment (xuhdev) [inline tools/setup_helpers/cuda.py]: You can use the `escape_path` utility here

Comment (peterjc123) [reply tools/setup_helpers/cuda.py]: Done!

Comment (xuhdev) [reply tools/setup_helpers/cuda.py]: I should have added that since you use `escape_path`, the `if IS_WINDOWS:` line is no longer necessary (:

Comment (peterjc123) [reply tools/setup_helpers/cuda.py]: No problem, fixed.

**Comment by mrshenli** [comment]
Hi @peterjc123, could you please add some more context to the PR description, e.g., what caused the error previously and how it is fix. This will help us better triage and document this PR in the release note. Thanks!

**Comment by peterjc123** [comment]
@mrshenli Okay, the issue is triggered when someone has no CUDA installed (or not installed in the default locations). Previously, we will try to call `which nvcc` to find the executable in `PATH`, and then infer `CUDA_HOME` from that result. But using `subprocess` sometimes throws exceptions and we didn't guard this previously. So we changed that to use the `which` function in `cmake.py`, which won't raise exceptions if things are not found. In order to reuse that, the code refactoring is made to move that function into `__init__.py`.

**Review by xuhdev — APPROVED** [review]

**Pull request closed without merging**